src/main/java/org/broadinstitute/hellbender/utils/io/TextFeatureReader.java

```diff
@@ -0,0 +1,244 @@
+import java.util.NoSuchElementException;
+
+public final class TextFeatureReader<F extends Feature> implements FeatureReader<F> {
+    private final Path path;
+    private final FeatureCodec<F, Iterator<String>> codec;
+    private FeatureCodecHeader header;
+    private Boolean queryable;
+    private Index index;
+
+    private static final long BOGUS_FILE_OFFSET = -1L;
+
+    public TextFeatureReader( final GATKPath path,
+                              final FeatureCodec<F, Iterator<String>> codec ) {
+        this.path = path.toPath();
+        this.codec = codec;
+    }
+
+    @Override
+    public FeatureIterator<F> iterator() {
+        return new FeatureIterator<>(path, codec);
+    }
+
+    @Override
+    public boolean isQueryable() {
+        if ( queryable == null ) {
+            try {
+                queryable = AbstractFeatureReader.isTabix(path.toString(), null);
+            } catch ( final IOException ioe ) {
+                throw new UserException("Error while checking for existence of index for " + path);
+            }
+        }
+        return queryable;
+    }
+
+    @Override
+    public FeatureIterator<F> query( final String chr, final int start, final int end ) {
+        if ( index == null ) {
+            if ( !isQueryable() ) {
+                throw new UserException("Querying non-queryable feature reader for " + path);
+            }
+            index = IndexUtils.loadTabixIndex(path);
+            if ( index == null ) {
+                throw new UserException("Despite its appearing to exist, couldn't load index for " + path);
+            }
+        }
+        final List<Block> blocks = index.getBlocks(chr, start, end);
+        return new FeatureIterator<>(path,
+                codec,
+                new SimpleInterval(chr, start, end),
+                blocks == null || blocks.isEmpty() ? BOGUS_FILE_OFFSET : blocks.get(0).getStartPosition());
+    }
+
+    @Override
+    public void close() {}
+
+    @Override
+    public List<String> getSequenceNames() {
+        return Collections.emptyList();
+    }
+
+    @Override
+    public Object getHeader() {
+        if ( header == null ) {
+            final LineIterator lineIterator = new LineIterator(path);
+            try {
+                header = codec.readHeader(lineIterator);
+            } catch ( final IOException ioe ) {
+                throw new UserException("Can't read header from " + path, ioe);
+            }
+            lineIterator.close();
+        }
+        return header.getHeaderValue();
+    }
+
+    private final static class LineIterator implements Iterator<String> {
+        private final Path path;
+        private final BufferedReader reader;
+        private String nextLine;
+
+        public LineIterator( final Path path ) {
+            this.path = path;
+            try {
+                InputStream in = new BufferedInputStream(Files.newInputStream(path), 256*1024);
+                if ( path.toString().endsWith(".gz") ) {
```

Comment:
It's safer to use htsjdk's IOUtil.isBlockCompressed() because it handles things like http query parameter stripping.